src/pounder/mod.rs

```diff
@@ -0,0 +1,186 @@
+                               hal::gpio::gpiog::PG7<hal::gpio::Output<hal::gpio::PushPull>>>,
+    mcp23017: mcp23017::MCP23017<hal::i2c::I2c<hal::stm32::I2C1>>,
+    attenuator_spi: hal::spi::Spi<hal::stm32::SPI1>,
+    adc1: hal::adc::Adc<hal::stm32::ADC1, hal::adc::Enabled>,
+    adc2: hal::adc::Adc<hal::stm32::ADC2, hal::adc::Enabled>,
+    adc1_in_p: hal::gpio::gpiof::PF11<hal::gpio::Analog>,
+    adc2_in_p: hal::gpio::gpiof::PF14<hal::gpio::Analog>,
+}
+
+impl<DELAY> PounderDevices<DELAY>
+where
+    DELAY: embedded_hal::blocking::delay::DelayMs<u8>,
+{
+    pub fn new(mcp23017: mcp23017::MCP23017<hal::i2c::I2c<hal::stm32::I2C1>>,
+               ad9959: ad9959::Ad9959<QspiInterface,
+                                      DELAY,
+                                      hal::gpio::gpiog::PG7<
+                                        hal::gpio::Output<hal::gpio::PushPull>>>,
+               attenuator_spi: hal::spi::Spi<hal::stm32::SPI1>,
+               adc1: hal::adc::Adc<hal::stm32::ADC1, hal::adc::Enabled>,
+               adc2: hal::adc::Adc<hal::stm32::ADC2, hal::adc::Enabled>,
+               adc1_in_p: hal::gpio::gpiof::PF11<hal::gpio::Analog>,
+               adc2_in_p: hal::gpio::gpiof::PF14<hal::gpio::Analog>,
+               ) -> Result<Self, Error> {
+        let mut devices = Self {
+            mcp23017,
+            ad9959,
+            attenuator_spi,
+            adc1,
+            adc2,
+            adc1_in_p,
+            adc2_in_p,
+        };
+
+        // Configure power-on-default state for pounder. All LEDs are on, on-board oscillator
+        // selected, attenuators out of reset.
+        devices.mcp23017.write_gpioa(0xF).map_err(|_| Error::I2c)?;
+        devices.mcp23017.write_gpiob(1_u8.wrapping_shl(5)).map_err(|_| Error::I2c)?;
+        devices.mcp23017.all_pin_mode(mcp23017::PinMode::OUTPUT).map_err(|_| Error::I2c)?;
+
+        devices.select_onboard_clock()?;
+
+        Ok(devices)
+    }
+
+    pub fn select_external_clock(&mut self, frequency: u32) -> Result<(), Error>{
+        self.mcp23017.digital_write(EXT_CLK_SEL_PIN, 1).map_err(|_| Error::I2c)?;
+        self.ad9959.set_clock_frequency(frequency).map_err(|_| Error::DDS)?;
+
+        Ok(())
+    }
+
+    pub fn select_onboard_clock(&mut self) -> Result<(), Error> {
+        self.mcp23017.digital_write(EXT_CLK_SEL_PIN, 0).map_err(|_| Error::I2c)?;
+        self.ad9959.set_clock_frequency(100_000_000).map_err(|_| Error::DDS)?;
+
+        Ok(())
+    }
+}
+
+impl<DELAY> AttenuatorInterface for PounderDevices<DELAY>
+{
+    fn reset(&mut self) -> Result<(), Error> {
+        self.mcp23017.digital_write(ATT_RST_N_PIN, 1).map_err(|_| Error::I2c)?;
+        // TODO: Delay here.
+        self.mcp23017.digital_write(ATT_RST_N_PIN, 0).map_err(|_| Error::I2c)?;
+
+        Ok(())
+    }
+
+    fn latch(&mut self, channel: DdsChannel) -> Result<(), Error> {
+        let pin = match channel {
+            DdsChannel::Zero => ATT_LE0_PIN,
```

Comment:
The mapping attenuator to dds channel is not 1:1. DDS 0,1,2,3 correspond to Attenuators 1,0,3,2. Attenuator byte order in the daisy chain and latch mapping is 0,1,2,3.